risc0/zkp/rust/src/field/goldilocks.rs

```diff
@@ -0,0 +1,745 @@
+// Copyright 2022 Risc0, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/// ! Goldilocks field.
+/// ! Support for the base finite field modulo 2^64 - 2^32 + 1
+use crate::field::{self, Elem as FieldElem};
+
+use core::ops;
+
+use bytemuck::{Pod, Zeroable};
+
+/// The Goldilocks class is an element of the finite field F_p, where P is the
+/// prime number 2^64 - 2^32 + 1. Here we implement integer
+/// arithmetic modulo P for both Goldilocks and for a field extension of
+/// Goldilocks.
+///
+/// The `Fp` datatype is the core type of all of the operations done within the
+/// zero knowledge proofs, and is the smallest 'addressable' datatype, and the
+/// base type of which all composite types are built. In many ways, one can
+/// imagine it as the word size of a strange architecture,
+/// and its operations as wrapping operations which respect word size P.
+///
+/// The Fp class wraps all standard arithmetic operations to make finite
+/// field elements appear like ordinary numbers (which, for the most part, they
+/// are).
+#[derive(Eq, PartialEq, Clone, Copy, Debug, Pod, Zeroable)]
+#[repr(transparent)]
+pub struct Elem(u64);
+
+impl Default for Elem {
+    /// As a default, return the zero [Elem].
+    fn default() -> Self {
+        Self::ZERO
+    }
+}
+
+/// The modulus of our Goldilocks field: 2^64 - 2^32 + 1
+const P: u64 = (0u64.wrapping_sub(1) << 32) + 1;
```

Comment:
The way this is computed maybe warrants a mild explaination?  Basically, the obvious thing of computing (1 << 64) - (1 << 32) + 1 would overflow U64, so we need to be trixy.